part_two.js

```diff
@@ -1 +1,164 @@
+var​ store = { 
+  report: {
+    4​: { id: ​4​, title: ​'Sample Report'​ },
+    21​: { id: ​21​, title: ​'Portfolio Summary 2020'​ },
+    }, 
+  document: {
+    8​: { id: ​8​, report_id: ​4​, name: ​'Sample Document'​, filetype: ​'txt'​ },
+    34​: { id: ​34​, report_id: ​21​, name: ​'Quarterly Report'​, filetype: ​'pdf'​ }, 
+    87​: { id: ​87​, report_id: ​21​, name: ​'Performance Summary'​, filetype: ​'pdf'​}
+    }, 
+  page: {
+    19​: { id: ​19​, document_id: ​34, body: ​'Lorem ipsum...'​, footnote: ​null​ },
+    21: { id: ​21, document_id: ​8, body: ​'Lorem ipsum...'​, footnote: ​null​ },
+    72​: { id: ​72​, document_id: ​87​, body: ​'Ut aliquet...'​, footnote: ​'Aliquam erat...'​ }, 
+    205​: { id: ​205​, document_id: ​34​, body: ​'Donec a dui et...'​, footnote: ​null​ },
+  },
+}
+
+
+
+/*  Prompt 1 */
+
+function mapReport(store){
+  var result, report, reports, reportId, documents, pages, id;
+
+  reports = store.report;
+  documents = store.document;
+  pages = store.page;
+  result = {};
+
+  // this is assuming that there are checks upstream to ensure clean store val (ie, no dulipcates in this map)
+  for (id in pages){
+    report = reports[documents[pages[id]?.document_id]?.report_id]
+    if (report != undefined){
+      reportId = report.id
+      if (result[reportId]) {
+          ++result[reportId].number_of_pages
+      } else {
+          result[reportId]= {'report_title': report.title, 'report_id': reportId, 'number_of_pages': 1}
+      }
+    }
+  }
+  return result;
+}
+
+/* example function call:
+  mapReport(store)
+ --------------------- */
+
+/* Prompt 2 */
+  
+function mapStore (store){
+  var reports, documents, pages, parentReports;
+
+  reports = store.report;
+  documents = store.document;
+  pages = store.page;
+
+  parentReports = mapObjectsToParentReport( mapObjectsToParentReport(pages, documents, 'document_id'), reports, 'report_id');
+  return (Object.values(parentReports));
+}
```

Comment:
probably could have combined mapStore with mapObjectsToParentReport into some sort of recursive function, but recursive functions are harder to debug and harder to read, I prefer to split things up more. Also I'm still traumatized by the recursion problems in CS finals.